core/src/stdlib/pyscript/events.py

```diff
@@ -1,134 +1,224 @@
 import inspect
-import sys
-
 from functools import wraps
-from pyscript.magic_js import document
+from pyscript.context import document
 from pyscript.ffi import create_proxy
 from pyscript.util import is_awaitable
-from pyscript import config
 
 
 class Event:
     """
-    Represents something that may happen at some point in the future.
+    A custom event that can notify multiple listeners when triggered.
+
+    Use this class to create your own event system within Python code.
+    Listeners can be either regular functions or async functions.
+
+    ```python
+    from pyscript.events import Event
+
+    # Create a custom event.
+    data_loaded = Event()
+
+    # Add a listener.
+    def on_data_loaded(result):
+        print(f"Data loaded: {result}")
+
+    data_loaded.add_listener(on_data_loaded)
+
+    # Time passes.... trigger the event.
+    data_loaded.trigger({"data": 123})
+    ```
     """
 
     def __init__(self):
         self._listeners = []
 
     def trigger(self, result):
         """
-        Trigger the event with a result to pass into the handlers.
+        Trigger the event and notify all listeners with the given `result`.
         """
         for listener in self._listeners:
             if is_awaitable(listener):
-                # Use create task to avoid making this an async function.
                 asyncio.create_task(listener(result))
             else:
                 listener(result)
 
     def add_listener(self, listener):
         """
-        Add a callable/awaitable to listen to when this event is triggered.
+        Add a function to be called when this event is triggered.
+
+        The `listener` must be callable. It can be either a regular function
+        or an async function. Duplicate listeners are ignored.
         """
-        if is_awaitable(listener) or callable(listener):
-            if listener not in self._listeners:
-                self._listeners.append(listener)
-        else:
-            msg = "Listener must be callable or awaitable."
+        if not callable(listener):
+            msg = "Listener must be callable."
             raise ValueError(msg)
+        if listener not in self._listeners:
+            self._listeners.append(listener)
 
-    def remove_listener(self, *args):
+    def remove_listener(self, *listeners):
         """
-        Clear the specified handler functions in *args. If no handlers
-        provided, clear all handlers.
+        Remove specified `listeners`. If none specified, remove all listeners.
         """
-        if args:
-            for listener in args:
-                self._listeners.remove(listener)
+        if listeners:
+            for listener in listeners:
+                try:
+                    self._listeners.remove(listener)
+                except ValueError:
+                    pass  # Silently ignore listeners not in the list.
         else:
             self._listeners = []
 
 
-def when(target, *args, **kwargs):
```

Comment:
IIRC this was to be able to also pass `{"once": True}` or any other DOM option one could pass along: https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#options

removing `**kwargs` works with current tests and use cases but it blocks any further development or ability to also pass specific option fields that can also be **mandatory** to have listeners work such as `passive`, `capture` or even `signal` ... can this be an opportunity to instead improve this `@when` decorator?

Reply:
P.S. my hint is just to add `options=None` as third argument, that should guarantee present and future compatibility with DOM APIs.

Reply:
OK. So I like your `options` approach but used `**options` so we capture them as a dict of args. These are just added to the `addEventListener` call, and if no `**options` are passed then there's no side-effects (because an empty option is treated as no args). 

I've added a couple of tests:

1.  To show that this functionality works, and,
2.  It's also ignored in the case of an `Event` based `event_type`. 

:+1: